src/feature/hs/hs_common.c

```diff
@@ -1723,45 +1716,41 @@ hs_get_extend_info_from_lspecs(const smartlist_t *lspecs,
     }
   } SMARTLIST_FOREACH_END(ls);
 
-  /* Legacy ID is mandatory, and we require IPv4. */
-  if (!have_v4 || !have_legacy_id) {
+  /* Choose a preferred address first, but fall back to an allowed address.
+   * Skip IPv6 if we're not direct as relays only extend via IPv4. */
+  fascist_firewall_choose_address_ls(lspecs, 0, &ap, direct_conn);
```

Comment:
Don't call fascist_firewall_choose_address_ls() when direct_conn is false.
Instead, use the IPv4 address, or return NULL if there is no IPv4 address.

Reply:
The function will return NULL if you say:
```
goto done;
```